File: ToC_2026/8/question4/Java/4.java
import java.util.Scanner;

public class Q4_Mod {
    static boolean iszero(int m) { return m == 0; }
    static int pi(int m) { return m == 0 ? 0 : m - 1; }

    static boolean lt(int m, int n) {
        if (iszero(n)) return false;
        else if (iszero(m)) return true;
        else return lt(pi(m), pi(n));
    }

    static int cut_off_sub(int m, int n) {
        if (iszero(n)) return m;
        else return cut_off_sub(pi(m), pi(n));
    }

    static int mod(int m, int n) {
        if (iszero(n)) {
            return -1;
        }
        if (lt(m, n)) {
            return m;
        } else {
            return mod(cut_off_sub(m, n), n);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter two numbers (m n): ");
        int m = scanner.nextInt();
        int n = scanner.nextInt();
        int result = mod(m, n);
        if (result == -1) {
            System.out.println("Error: Cannot divide by zero.");
        } else {
            System.out.println("The remainder of " + m + " divided by " + n + " is " + result);
        }
    }
}

File: ToC_2026/8/question4/Rust/4.rs
use std::io;

fn iszero(m: i32) -> bool { m == 0 }
fn pi(m: i32) -> i32 { if m == 0 { 0 } else { m - 1 } }

fn lt(m: i32, n: i32) -> bool {
    if iszero(n) { false }
    else if iszero(m) { true }
    else { lt(pi(m), pi(n)) }
}

fn cut_off_sub(m: i32, n: i32) -> i32 {
    if iszero(n) { m } else { cut_off_sub(pi(m), pi(n)) }
}

fn mod_func(m: i32, n: i32) -> i32 {
    if iszero(n) {
        -1
    } else if lt(m, n) {
        m
    } else {
        mod_func(cut_off_sub(m, n), n)
    }
}

fn main() {
    println!("Enter two numbers (m n): ");
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let nums: Vec<i32> = input.split_whitespace().map(|s| s.parse().unwrap()).collect();
    
    let result = mod_func(nums[0], nums[1]);
    if result == -1 {
        println!("Error: Cannot divide by zero.");
    } else {
        println!("The remainder of {} divided by {} is {}", nums[0], nums[1], result);
    }
}